src/node_jsvmapi.cc

```diff
@@ -1569,122 +1600,153 @@ napi_status napi_coerce_to_string(napi_env e, napi_value v, napi_value* result)
+  assert(obj->InternalFieldCount() > 0);
+  obj->SetAlignedPointerInInternalField(0, nativeObj);
 
-  if (handle)
-  {
-    return napi_create_weakref(
-      e,
-      v8impl::JsValueFromV8LocalValue(wrap->handle()),
-      handle);
-  }
+  v8impl::Reference* reference = new v8impl::Reference(
+    isolate, obj, 0, false, finalize_cb, nativeObj);
+  *result = reinterpret_cast<napi_ref>(reference);
 
-  // TODO: Is the handle parameter really optional?
-  //       Why would anyone want to construct an object wrap and immediately lose it?
   return GET_RETURN_STATUS();
 }
 
 napi_status napi_unwrap(napi_env e, napi_value jsObject, void** result) {
   NAPI_PREAMBLE(e);
+  CHECK_ARG(jsObject);
   CHECK_ARG(result);
 
-  v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
-  v8::Local<v8::Context> context = isolate->GetCurrentContext();
-  v8::Local<v8::Object> obj;
-  CHECK_TO_OBJECT(context, obj, jsObject);
+  v8::Local<v8::Object> obj = v8impl::V8LocalValueFromJsValue(jsObject).As<v8::Object>();
+
+  assert(obj->InternalFieldCount() > 0);
+  *result = obj->GetAlignedPointerFromInternalField(0);
 
-  *result = v8impl::ObjectWrapWrapper::Unwrap(obj);
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_create_persistent(napi_env e, napi_value v, napi_persistent* result) {
+napi_status napi_create_external(napi_env e,
+                                 void* data,
+                                 napi_finalize finalize_cb,
+                                 napi_value* result) {
   NAPI_PREAMBLE(e);
   CHECK_ARG(result);
 
   v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
-  v8::Persistent<v8::Value> *thePersistent =
-      new v8::Persistent<v8::Value>(
-          isolate, v8impl::V8LocalValueFromJsValue(v));
 
-  *result = v8impl::JsPersistentFromV8PersistentValue(thePersistent);
+  v8::Local<v8::Value> externalValue = v8::External::New(isolate, data);
+
+  // The Reference object will delete itself after invoking the finalizer callback.
+  new v8impl::Reference(isolate, externalValue, 0, true, finalize_cb, data);
+
+  *result = v8impl::JsValueFromV8LocalValue(externalValue);
+
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_release_persistent(napi_env e, napi_persistent p) {
+napi_status napi_get_value_external(napi_env e, napi_value v, void** result) {
   NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
 
-  v8::Persistent<v8::Value> *thePersistent =
-      v8impl::V8PersistentValueFromJsPersistentValue(p);
-  thePersistent->Reset();
-  delete thePersistent;
+  v8::Local<v8::External> externalValue = v8impl::V8LocalValueFromJsValue(v).As<v8::External>();
+  *result = externalValue->Value();
 
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_get_persistent_value(napi_env e, napi_persistent p, napi_value* result) {
+
+// Set initial_refcount to 0 for a weak reference, >0 for a strong reference.
+// The finalizer callback and callback data parameters optional.
```

Comment:
JsCreateWeakReference is itself a new JSRT API, if I am remembering correctly. It hasn't flowed-back into JSRT-proper so we are free to work on it. It sounds like we should just add support for a finalizer callback. I can get behind the need for this.